grails-app/services/org/pih/warehouse/data/LocationDataService.groovy

```diff
@@ -112,15 +123,23 @@ class LocationDataService {
             location = new Location()
         }
 
+        def locationTypeToBeChanged = params.locationType ? LocationType.findByNameLike(params.locationType + "%") : null
+        def currentLocationType = location.locationType
+
         location.name = params.name
         location.active = Boolean.valueOf(params.active)
         location.locationNumber = params.locationNumber
-        location.locationType = params.locationType ? LocationType.findByNameLike(params.locationType + "%") : null
         location.locationGroup = params.locationGroup ? LocationGroup.findByName(params.locationGroup) : null
         location.parentLocation = params.parentLocation ? Location.findByNameOrLocationNumber(params.parentLocation, params.parentLocation) : null
         location.organization = params.organization ? Organization.findByCodeOrName(params.organization, params.organization) : null
         location.address = createOrUpdateAddress(params, location?.address?.id)
 
+        //Do not allow to change internal type location to non-internal
+        if (((currentLocationType?.isInternalLocation() || currentLocationType?.isZone()) && (locationTypeToBeChanged?.isInternalLocation() || locationTypeToBeChanged?.isZone())) || !(currentLocationType?.isInternalLocation() || currentLocationType?.isZone())) {
+            location.locationType = locationTypeToBeChanged
+            location.supportedActivities = currentLocationType && locationTypeToBeChanged.equals(currentLocationType) ? location.supportedActivities : new HashSet(location.locationType.getSupportedActivities())
```

Comment:
If you want to add something to the supported activities, you should do `location.addToSupportedActivities(...)`, but is there a need to add something to the location's supported activities if it's already on the location type?

Reply:
It occures that using `location.addToSupportedActivities(...)` is not possible because of lazy initialization

Reply:
That probably just means that the location has validation errors (i.e. location.validate() is false) before we add to supported activities. But I would be really careful about changing the supported activities here. Is this really a requirement? I think only superusers should be able to override the supported activities of the location. It feels like the LocationType import should be responsible for that.